src/Pos/Generator/BlockEvent.hs

```diff
@@ -41,13 +94,388 @@ data BlockRollbackResult
                                 * genesis block rollback
                                 * etc -}
 
-data BlockEventRollback = BlockEventRollback
-    { _berInput    :: ![Blund SscGodTossing]
+instance IsBlockEventFailure BlockRollbackResult where
+    isBlockEventFailure = \case
+        BlockRollbackFailure -> True
+        _ -> False
+
+data BlockEventRollback' blund = BlockEventRollback
+    { _berInput    :: !(NewestFirst NE blund)
     , _berOutValid :: !BlockRollbackResult
+    } deriving (Functor, Foldable)
+
+makeLenses ''BlockEventRollback'
+
+instance IsBlockEventFailure (BlockEventRollback' blund) where
+    isBlockEventFailure = isBlockEventFailure . view berOutValid
+
+type BlockEventRollback = BlockEventRollback' BlundDefault
+
+data BlockEvent' blund
+    = BlkEvApply (BlockEventApply' blund)
+    | BlkEvRollback (BlockEventRollback' blund)
+    deriving (Functor, Foldable)
+
+instance IsBlockEventFailure (BlockEvent' blund) where
+    isBlockEventFailure = \case
+        BlkEvApply    a -> isBlockEventFailure a
+        BlkEvRollback a -> isBlockEventFailure a
+
+type BlockEvent = BlockEvent' BlundDefault
+
+----------------------------------------------------------------------------
+-- Block event generation
+----------------------------------------------------------------------------
+
+{- |
+  Block indices. They're neither 0-based nor 1-based, they can start from any
+  index (even negative). Therefore it's necessary to adjust them before using
+  for lookup in a data structure.
+-}
+newtype BlockIndex = BlockIndex {getBlockIndex :: Int}
+    deriving (Eq, Ord, Num, Real, Integral, Enum, Read, Show,
+              Buildable, Generic, Typeable, NFData, Hashable, Random)
+
+newtype BlockEventCount = BlockEventCount {getBlockEventCount :: Word64}
+    deriving (Eq, Ord, Num, Real, Integral, Enum, Read, Show,
+              Buildable, Generic, Typeable, NFData, Hashable, Random)
+
+-- | A coefficient in the range [0,1]. Pass it to 'weighted' if you ever get
+-- the chance.
+newtype Chance = Chance {getChance :: Rational}
+
+-- | Generate a boolean that may happen to be of true value.
+byChance :: (Monad m, RandomGen g) => Chance -> RandT g m Bool
+byChance (Chance c) = weighted [(False, 1 - c), (True, c)]
+
+data BlockEventGenParams = BlockEventGenParams
+    { _begpSecrets         :: !AllSecrets
+    , _begpBlockCountMax   :: !BlockCount {- ^ the maximum possible amount of
+    blocks in a BlockApply event. Must be 1 or more. Can be violated by 1 in
+    some cases (see Note on reordering corner cases), so if you specify '52'
+    here, some rare events may contain up to '53' blocks. -}
+    , _begpBlockEventCount :: !BlockEventCount
```

Comment:
Can you comment this field too? AFAIU, it's the **maximal** number of events to generate. The actual number of events can be smaller. I think it's worth a comment, because otherwise someone can think that it **always** generates as many events.